File: numbers/easy/java_solutions/HappyNumber_202.java
package numbers.easy.java_solutions;

import java.util.HashSet;
import java.util.Set;

public class HappyNumber_202 {

    public boolean isHappy(int n) {
        Set<Integer> set = new HashSet<>();

        while (n != 1) {
            var nextNumber = 0;
            set.add(n);

            while (n != 0) {
                var rightNumber = n % 10;
                nextNumber += rightNumber * rightNumber;
                n /= 10;
            }
            if (set.contains(nextNumber)) {
                return false;
            }
            n = nextNumber;
        }
        return true;
    }
}

File: leetcode/numbers/easy/golang_solutions/happy_number_202.go
package golang_solutions

func isHappy(n int) bool {
	set := make(map[int]interface{})
	for n != 1 {
		nextNumber := 0
		set[n] = nil
		for n != 0 {
			right := n % 10
			nextNumber += right * right
			n /= 10
		}
		if _, ok := set[nextNumber]; ok {
			return false
		}
		n = nextNumber
	}
	return true
}
